File: android_companion/lib/services/sync_merger.dart
import 'dart:async';
import 'package:sqflite/sqflite.dart';

/// Conflict resolution options
enum ConflictResolution {
  keepLocal,
  keepRemote,
  alwaysLocal,
  alwaysRemote,
  cancel,
}

/// Information about a sync conflict
class ConflictInfo {
  final String table;
  final String uuid;
  final String localModified;
  final String remoteModified;
  final Map<String, dynamic> localData;
  final Map<String, dynamic> remoteData;
  final String description;

  ConflictInfo({
    required this.table,
    required this.uuid,
    required this.localModified,
    required this.remoteModified,
    required this.localData,
    required this.remoteData,
    required this.description,
  });
}

/// Callback type for conflict resolution
typedef ConflictCallback =
    Future<ConflictResolution> Function(ConflictInfo conflict);

/// Tables that support sync
const List<String> syncableTables = [
  'decks',
  'flashcards',
  'imported_content',
  'word_definitions',
  'sentence_explanations',
  'writing_sessions',
  'chat_sessions',
  'grammar_book_entries',
  'collections',
];

/// Sync Merger - handles row-by-row comparison between databases
class SyncMerger {
  final Database localDb;
  final Database remoteDb;
  final String lastSyncTime;

  ConflictResolution? _alwaysPreference;
  ConflictCallback? onConflict;

  SyncMerger({
    required this.localDb,
    required this.remoteDb,
    required this.lastSyncTime,
  });

  Future<Map<String, dynamic>> _getAllRows(Database db, String table) async {
    try {
      final List<Map<String, dynamic>> rows = await db.query(
        table,
        where: 'deleted_at IS NULL',
      );
      final Map<String, dynamic> result = {};
      for (final row in rows) {
        if (row['uuid'] != null) {
          result[row['uuid'] as String] = Map<String, dynamic>.from(row);
        }
      }
      return result;
    } catch (e) {
      return {};
    }
  }

  Future<Set<String>> _getDeletedUuids(Database db, String table) async {
    try {
      final List<Map<String, dynamic>> rows = await db.query(
        table,
        columns: ['uuid'],
        where: 'deleted_at IS NOT NULL',
      );
      return rows
          .where((r) => r['uuid'] != null)
          .map((r) => r['uuid'] as String)
          .toSet();
    } catch (e) {
      return {};
    }
  }

  bool _isModifiedSinceSync(String? lastModified) {
    if (lastModified == null || lastModified.isEmpty) return false;
    return lastModified.compareTo(lastSyncTime) > 0;
  }

  Map<String, dynamic> _resolveSrsConflict(
    Map<String, dynamic> local,
    Map<String, dynamic> remote,
  ) {
    final localReviews = (local['total_reviews'] as int?) ?? 0;
    final remoteReviews = (remote['total_reviews'] as int?) ?? 0;
    return localReviews >= remoteReviews ? local : remote;
  }

  String _getItemDescription(String table, Map<String, dynamic> row) {
    switch (table) {
      case 'flashcards':
        return 'Flashcard: ${(row['question'] as String? ?? 'Unknown').substring(0, 50.clamp(0, (row['question'] as String? ?? '').length))}';
      case 'decks':
        return 'Deck: ${row['name'] ?? 'Unknown'}';
      case 'imported_content':
        return 'Import: ${(row['content'] as String? ?? 'Unknown').substring(0, 50.clamp(0, (row['content'] as String? ?? '').length))}';
      case 'word_definitions':
        return 'Word: ${row['word'] ?? 'Unknown'}';
      default:
        return '$table: ${(row['uuid'] as String? ?? 'Unknown').substring(0, 8)}';
    }
  }

  Future<(int, int, int)> mergeTable(String table) async {
    final localRows = await _getAllRows(localDb, table);
    final remoteRows = await _getAllRows(remoteDb, table);
    final localDeleted = await _getDeletedUuids(localDb, table);
    final remoteDeleted = await _getDeletedUuids(remoteDb, table);

    int added = 0;
    int updated = 0;
    int conflicts = 0;

    for (final entry in remoteRows.entries) {
      final uuid = entry.key;
      final remoteRow = entry.value;

      if (localDeleted.contains(uuid)) continue;

      final localRow = localRows[uuid];

      if (localRow == null) {
        // New item from remote
        await _insertRow(table, remoteRow);
        added++;
      } else {
        final localModified = _isModifiedSinceSync(
          localRow['last_modified'] as String?,
        );
        final remoteModified = _isModifiedSinceSync(
          remoteRow['last_modified'] as String?,
        );

        if (localModified && remoteModified) {
          conflicts++;

          // Flashcard SRS: auto-resolve
          if (table == 'flashcards') {
            final winner = _resolveSrsConflict(localRow, remoteRow);
            if (winner == remoteRow) {
              await _updateRow(table, remoteRow);
              updated++;
            }
            continue;
          }

          // Check session preference
          if (_alwaysPreference == ConflictResolution.alwaysLocal) {
            continue;
          } else if (_alwaysPreference == ConflictResolution.alwaysRemote) {
            await _updateRow(table, remoteRow);
            updated++;
            continue;
          }

          // Ask user via callback
          if (onConflict != null) {
            final conflict = ConflictInfo(
              table: table,
              uuid: uuid,
              localModified: localRow['last_modified'] ?? '',
              remoteModified: remoteRow['last_modified'] ?? '',
              localData: localRow,
              remoteData: remoteRow,
              description: _getItemDescription(table, localRow),
            );

            final resolution = await onConflict!(conflict);

            if (resolution == ConflictResolution.cancel) {
              throw Exception('Sync cancelled by user');
            } else if (resolution == ConflictResolution.keepRemote ||
                resolution == ConflictResolution.alwaysRemote) {
              await _updateRow(table, remoteRow);
              updated++;
              if (resolution == ConflictResolution.alwaysRemote) {
                _alwaysPreference = ConflictResolution.alwaysRemote;
              }
            } else if (resolution == ConflictResolution.alwaysLocal) {
              _alwaysPreference = ConflictResolution.alwaysLocal;
            }
          }
        } else if (remoteModified && !localModified) {
          await _updateRow(table, remoteRow);
          updated++;
        }
      }
    }

    // Handle remote deletions
    for (final uuid in remoteDeleted) {
      if (localRows.containsKey(uuid)) {
        final localRow = localRows[uuid]!;
        final localModified = _isModifiedSinceSync(
          localRow['last_modified'] as String?,
        );
        if (!localModified) {
          await _softDeleteRow(table, uuid);
        }
      }
    }

    return (added, updated, conflicts);
  }

  Future<void> _insertRow(String table, Map<String, dynamic> row) async {
    final columns = row.keys.where((k) => k != 'id').toList();
    final values = columns.map((c) => row[c]).toList();
    final placeholders = columns.map((_) => '?').join(', ');
    final colNames = columns.join(', ');
    await localDb.rawInsert(
      'INSERT INTO $table ($colNames) VALUES ($placeholders)',
      values,
    );
  }

  Future<void> _updateRow(String table, Map<String, dynamic> row) async {
    final columns = row.keys.where((k) => k != 'id' && k != 'uuid').toList();
    final setClause = columns.map((c) => '$c = ?').join(', ');
    final values = columns.map((c) => row[c]).toList();
    values.add(row['uuid']);
    await localDb.rawUpdate(
      'UPDATE $table SET $setClause WHERE uuid = ?',
      values,
    );
  }

  Future<void> _softDeleteRow(String table, String uuid) async {
    final now = DateTime.now().toIso8601String();
    await localDb.rawUpdate('UPDATE $table SET deleted_at = ? WHERE uuid = ?', [
      now,
      uuid,
    ]);
  }

  Future<Map<String, (int, int, int)>> mergeAll() async {
    final results = <String, (int, int, int)>{};
    for (final table in syncableTables) {
      try {
        results[table] = await mergeTable(table);
      } catch (e) {
        print('Error merging $table: $e');
        results[table] = (0, 0, 0);
      }
    }
    return results;
  }
}

File: src/services/sync_merger.py
"""
Sync Merger
-----------
Handles row-by-row comparison between local and remote databases.
Detects conflicts and provides resolution strategies.
"""

import sqlite3
from datetime import datetime
from typing import Dict, List, Tuple, Optional, Literal
from dataclasses import dataclass
from enum import Enum

class ConflictResolution(Enum):
    KEEP_LOCAL = "local"
    KEEP_REMOTE = "remote"
    ALWAYS_LOCAL = "always_local"
    ALWAYS_REMOTE = "always_remote"
    CANCEL = "cancel"

@dataclass
class ConflictInfo:
    table: str
    uuid: str
    local_modified: str
    remote_modified: str
    local_data: Dict
    remote_data: Dict
    description: str  # Human-readable item identifier

# Tables that support sync
SYNCABLE_TABLES = [
    "decks", "flashcards", "imported_content", "word_definitions",
    "sentence_explanations", "writing_sessions", "chat_sessions",
    "grammar_book_entries", "collections"
]

class SyncMerger:
    def __init__(self, local_db_path: str, remote_db_path: str, last_sync_time: Optional[str] = None):
        self.local_conn = sqlite3.connect(local_db_path)
        self.local_conn.row_factory = sqlite3.Row
        self.remote_conn = sqlite3.connect(remote_db_path)
        self.remote_conn.row_factory = sqlite3.Row
        self.last_sync_time = last_sync_time or "1970-01-01T00:00:00"
        
        # Session preferences
        self._always_preference: Optional[ConflictResolution] = None
        
        # Conflict handler callback (set by UI)
        self.on_conflict: Optional[callable] = None
    
    def close(self):
        self.local_conn.close()
        self.remote_conn.close()
    
    def _get_all_rows(self, conn: sqlite3.Connection, table: str) -> Dict[str, Dict]:
        """Get all rows from a table, keyed by UUID."""
        cursor = conn.cursor()
        try:
            cursor.execute(f"SELECT * FROM {table} WHERE deleted_at IS NULL")
            rows = cursor.fetchall()
            return {row['uuid']: dict(row) for row in rows if row['uuid']}
        except sqlite3.OperationalError:
            return {}  # Table might not exist
    
    def _get_deleted_uuids(self, conn: sqlite3.Connection, table: str) -> set:
        """Get UUIDs of soft-deleted rows."""
        cursor = conn.cursor()
        try:
            cursor.execute(f"SELECT uuid FROM {table} WHERE deleted_at IS NOT NULL")
            return {row[0] for row in cursor.fetchall() if row[0]}
        except sqlite3.OperationalError:
            return set()
    
    def _is_modified_since_sync(self, last_modified: str) -> bool:
        """Check if a row was modified after last sync."""
        if not last_modified:
            return False
        return last_modified > self.last_sync_time
    
    def _resolve_srs_conflict(self, local: Dict, remote: Dict) -> Dict:
        """Special handling for flashcard SRS data: most progress wins."""
        local_reviews = local.get('total_reviews', 0) or 0
        remote_reviews = remote.get('total_reviews', 0) or 0
        
        if local_reviews >= remote_reviews:
            return local
        return remote
    
    def _get_item_description(self, table: str, row: Dict) -> str:
        """Generate human-readable description for conflict dialog."""
        if table == "flashcards":
            return f"Flashcard: {row.get('question', 'Unknown')[:50]}"
        elif table == "decks":
            return f"Deck: {row.get('name', 'Unknown')}"
        elif table == "imported_content":
            return f"Import: {row.get('content', 'Unknown')[:50]}"
        elif table == "word_definitions":
            return f"Word: {row.get('word', 'Unknown')}"
        elif table == "sentence_explanations":
            return f"Sentence: {row.get('sentence', 'Unknown')[:50]}"
        elif table == "writing_sessions":
            return f"Essay: {row.get('topic', 'Unknown')[:50]}"
        elif table == "chat_sessions":
            return f"Chat: {row.get('cur_topic', 'Unknown')[:50]}"
        elif table == "grammar_book_entries":
            return f"Grammar: {row.get('title', 'Unknown')}"
        return f"{table}: {row.get('uuid', 'Unknown')[:8]}"
    
    def merge_table(self, table: str) -> Tuple[int, int, int]:
        """
        Merge a single table. Returns (added, updated, conflicts).
        """
        local_rows = self._get_all_rows(self.local_conn, table)
        remote_rows = self._get_all_rows(self.remote_conn, table)
        local_deleted = self._get_deleted_uuids(self.local_conn, table)
        remote_deleted = self._get_deleted_uuids(self.remote_conn, table)
        
        added = 0
        updated = 0
        conflicts = 0
        
        local_cursor = self.local_conn.cursor()
        
        # Process remote rows
        for uuid, remote_row in remote_rows.items():
            # Skip if deleted locally
            if uuid in local_deleted:
                continue
            
            local_row = local_rows.get(uuid)
            
            if local_row is None:
                # New item from remote - add it
                self._insert_row(local_cursor, table, remote_row)
                added += 1
            else:
                # Both exist - check for conflict
                local_modified = self._is_modified_since_sync(local_row.get('last_modified'))
                remote_modified = self._is_modified_since_sync(remote_row.get('last_modified'))
                
                if local_modified and remote_modified:
                    # CONFLICT!
                    conflicts += 1
                    
                    # Special case: Flashcard SRS - auto-resolve
                    if table == "flashcards":
                        winner = self._resolve_srs_conflict(local_row, remote_row)
                        if winner == remote_row:
                            self._update_row(local_cursor, table, remote_row)
                            updated += 1
                        continue
                    
                    # Check session preference
                    if self._always_preference == ConflictResolution.ALWAYS_LOCAL:
                        continue  # Keep local, do nothing
                    elif self._always_preference == ConflictResolution.ALWAYS_REMOTE:
                        self._update_row(local_cursor, table, remote_row)
                        updated += 1
                        continue
                    
                    # Ask user
                    if self.on_conflict:
                        conflict = ConflictInfo(
                            table=table,
                            uuid=uuid,
                            local_modified=local_row.get('last_modified', ''),
                            remote_modified=remote_row.get('last_modified', ''),
                            local_data=local_row,
                            remote_data=remote_row,
                            description=self._get_item_description(table, local_row)
                        )
                        resolution = self.on_conflict(conflict)
                        
                        if resolution == ConflictResolution.CANCEL:
                            raise Exception("Sync cancelled by user")
                        elif resolution in (ConflictResolution.KEEP_REMOTE, ConflictResolution.ALWAYS_REMOTE):
                            self._update_row(local_cursor, table, remote_row)
                            updated += 1
                            if resolution == ConflictResolution.ALWAYS_REMOTE:
                                self._always_preference = ConflictResolution.ALWAYS_REMOTE
                        elif resolution == ConflictResolution.ALWAYS_LOCAL:
                            self._always_preference = ConflictResolution.ALWAYS_LOCAL
                        # KEEP_LOCAL: do nothing
                
                elif remote_modified and not local_modified:
                    # Only remote changed - take remote
                    self._update_row(local_cursor, table, remote_row)
                    updated += 1
                # If only local changed or neither, keep local (do nothing)
        
        # Handle items deleted on remote
        for uuid in remote_deleted:
            if uuid in local_rows:
                local_row = local_rows[uuid]
                local_modified = self._is_modified_since_sync(local_row.get('last_modified'))
                
                if local_modified:
                    # Conflict: Remote deleted but local edited
                    # For now, keep local (user's edits win over deletions)
                    pass
                else:
                    # Safe to delete locally
                    self._soft_delete_row(local_cursor, table, uuid)
        
        self.local_conn.commit()
        return added, updated, conflicts
    
    def _insert_row(self, cursor, table: str, row: Dict):
        """Insert a new row."""
        columns = [k for k in row.keys() if k != 'id']
        placeholders = ', '.join(['?' for _ in columns])
        col_names = ', '.join(columns)
        values = [row[c] for c in columns]
        cursor.execute(f"INSERT INTO {table} ({col_names}) VALUES ({placeholders})", values)
    
    def _update_row(self, cursor, table: str, row: Dict):
        """Update existing row by UUID."""
        columns = [k for k in row.keys() if k not in ('id', 'uuid')]
        set_clause = ', '.join([f"{c} = ?" for c in columns])
        values = [row[c] for c in columns] + [row['uuid']]
        cursor.execute(f"UPDATE {table} SET {set_clause} WHERE uuid = ?", values)
    
    def _soft_delete_row(self, cursor, table: str, uuid: str):
        """Mark row as deleted."""
        now = datetime.now().isoformat()
        cursor.execute(f"UPDATE {table} SET deleted_at = ? WHERE uuid = ?", (now, uuid))
    
    def merge_all(self) -> Dict[str, Tuple[int, int, int]]:
        """Merge all syncable tables. Returns stats per table."""
        results = {}
        for table in SYNCABLE_TABLES:
            try:
                results[table] = self.merge_table(table)
            except Exception as e:
                print(f"Error merging {table}: {e}")
                results[table] = (0, 0, 0)
        return results
